iceoryx_examples/iceperf/iceoryx.cpp

```diff
@@ -68,24 +69,37 @@ void Iceoryx::shutdown() noexcept
 
 void Iceoryx::sendPerfTopic(uint32_t payloadSizeInBytes, bool runFlag) noexcept
 {
-    auto sendSample = static_cast<PerfTopic*>(m_publisher.allocateChunk(payloadSizeInBytes, true));
-    sendSample->payloadSize = payloadSizeInBytes;
-    sendSample->run = runFlag;
-    sendSample->subPackets = 1;
-
-    m_publisher.sendChunk(sendSample);
+    m_publisher.loan(payloadSizeInBytes).and_then([&](iox::popo::Sample<void>& sample) {
+        auto sendSample = static_cast<PerfTopic*>(sample.get());
+        sendSample->payloadSize = payloadSizeInBytes;
+        sendSample->run = runFlag;
+        sendSample->subPackets = 1;
+        sample.publish();
+    });
 }
 
 PerfTopic Iceoryx::receivePerfTopic() noexcept
 {
-    const void* receivedChunk;
-    while (!m_subscriber.getChunk(&receivedChunk))
+    const PerfTopic* receivedSample{nullptr};
+
+    do
     {
-        // poll as fast as possible
-    }
+        m_subscriber.take().and_then([&](iox::popo::Sample<const void>& sample) {
+            receivedSample = static_cast<const PerfTopic*>(sample.get());
```

Comment:
How do we prevent this? With the old API people were more aware that they should not copy pointers that are coming from the middleware. In ara::com it's a unique ptr to make this even harder. I guess our Sample is also just an advanced unique_ptr but when I look at this code I fear that people think they must call `sample.get()` to be able to work with the payload. But in fact `get()` is the most dangerous method. 

We need `get` for frameworks where you get a Sample from the iceoryx API and have to pass e.g. a void* to the next layer (like in ROS). Then this layer has to return the pointer and the sample must be released with the iceoryx API. 
How would this look like? @elBoberido @ithier 

- take a sample
- get the raw pointer with `sample.get()`
- release the sample for not releasing the chunk when it goes out of scope, `sample.release()`
- pass the raw pointer to the next layer
- something happens
- raw pointer is returned from the next layer
- create a Sample with the raw pointer
- let the Sample go out of scope for releasing the chunk  

Shoud we make an example for this? This is a use case a lot people have. Would it be better to have a `release()` that releases the ownersip but additionally returns the pointer like in the STL and to remove the get() method? Having a get() that gives you a pointer to a shared memory chunk which is released when the Sample goes out of scope is a dangerous thing . The STL unique_ptr also has it but maybe people are more aware there 

With the `Sample` class, the `get` method and the fact that the subscriber has no `releaseChunk` method any more, the lifetime and ownership question has some complexity. This must be well documented in the new API docu @elfenpiff @FerdinandSpitzschnueffler @MatthiasKillat

Reply:
This is a tough question and boils down to the raw pointer access like you pointed out. I'm not sure how to fix this, except making it harder to get to a non-owning raw pointer. e.g. by replacing `get` with `dangerousNonOwningRawPointer` and hope the user is too lazy to type that much, doesn't have auto-completion and is scared enough to try other things first.

The `std::unique_ptr` return a pointer with the `release` call, so this would be the solution when an owning pointer is necessary. For the non-owning use case this would open the door to chunk leaks, which we want to prevent in the first place. At first we need to know what the user wants to do with the raw pointer. If it's just storing for later use, then we have to prevent this and encourage to store the sample. If it's for copying, we should maybe enrich the `Sample` API, e.g.
```
T getPayloadCopy();
copyPayload(void* buffer, uint64_t bufferSize);
```
With this, we could get rid of the `get` method

Reply:
I would clearly state in the docu: Please use the typed API but if you are an expert, you know what you are doing and you want to get your hands dirty then use the untyped one. With this statement and the assumption that the users know what they are doing we could get rid of the `Sample` and offer an interface which looks like this:
```cpp
untypedSubscriber->take().and_then([&](const void * payload){
     auto header = freeFunctionToAcquireHeaderPtr(payload);
     doStuff(payload, header);
  
     untypedSubscriber->releaseRawSample(payload);
});
```

You could also pack this `payload` and `header` pointer into a `Sample` but then the user is maybe mislead to believe that the sample takes care of the release but when having a `void*` you clearly have to take care of this.

Maybe it is also then useful to not recommend the functional approach and stick to the classic approach and use it like:
```cpp
auto payload = untypedSubscriber->take();
if ( !payload.has_value() ) {
    doStuff(payload.value()); // acquires void *
    untypedSubscriber->releaseRawSample(*payload);
}
```

Reply:
moved this discussion to #408